File: bindings/c/tests/test_value.c
#include <float.h>
#include <math.h>
#include <setjmp.h>
#include <stdarg.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include <cmocka.h>

#include "cpublish.h"
#include "test_utils.h"

/*
----------------------------------------------------------------------------
  Checks
*/
static void test_cpublish_value_new_none_success(void **state) {
  CPublishStatus status;
  CPublishValue *value = cpublish_value_new_none();
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, &status), CPublishValueTypeNone);
  validate_status_ok(&status);

  cpublish_value_destroy(value);
  cpublish_status_destroy(&status);
}

void helper_test_cpublish_value_new_bool_success(bool v,
                                                 CPublishStatus *status) {
  CPublishValue *value = cpublish_value_new_bool(v);
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, status),
                   CPublishValueTypeBoolean);
  validate_status_ok(status);

  bool result = cpublish_value_bool(value, status);
  validate_status_ok(status);
  assert_int_equal(result, v);

  cpublish_value_destroy(value);
}

static void test_cpublish_value_new_bool_success(void **state) {
  CPublishStatus status;

  helper_test_cpublish_value_new_bool_success(true, &status);
  helper_test_cpublish_value_new_bool_success(false, &status);

  cpublish_status_destroy(&status);
}

void helper_test_cpublish_value_new_int_success(int64_t v,
                                                CPublishStatus *status) {
  CPublishValue *value = cpublish_value_new_int(v);
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, status),
                   CPublishValueTypeInteger);
  validate_status_ok(status);

  int64_t result = cpublish_value_int(value, status);
  validate_status_ok(status);
  assert_int_equal(result, v);

  cpublish_value_destroy(value);
}

static void test_cpublish_value_new_int_success(void **state) {
  CPublishStatus status;

  helper_test_cpublish_value_new_int_success(INT64_MIN, &status);
  helper_test_cpublish_value_new_int_success(-10, &status);
  helper_test_cpublish_value_new_int_success(-1, &status);
  helper_test_cpublish_value_new_int_success(0, &status);
  helper_test_cpublish_value_new_int_success(1, &status);
  helper_test_cpublish_value_new_int_success(10, &status);
  helper_test_cpublish_value_new_int_success(INT64_MAX, &status);

  cpublish_status_destroy(&status);
}

void helper_test_cpublish_value_new_float_success(double v,
                                                  CPublishStatus *status) {
  CPublishValue *value = cpublish_value_new_float(v);
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, status), CPublishValueTypeFloat);
  validate_status_ok(status);

  double result = cpublish_value_float(value, status);
  validate_status_ok(status);

  if (isnan(v)) {
    assert_true(isnan(result));
  } else {
    assert_double_equal(result, v, DBL_EPSILON);
  }

  cpublish_value_destroy(value);
}

static void test_cpublish_value_new_float_success(void **state) {
  CPublishStatus status;

  helper_test_cpublish_value_new_float_success(DBL_MIN, &status);
  helper_test_cpublish_value_new_float_success(-10.0, &status);
  helper_test_cpublish_value_new_float_success(-1.0, &status);
  helper_test_cpublish_value_new_float_success(0.0, &status);
  helper_test_cpublish_value_new_float_success(1.0, &status);
  helper_test_cpublish_value_new_float_success(10.0, &status);
  helper_test_cpublish_value_new_float_success(DBL_MAX, &status);
  helper_test_cpublish_value_new_float_success(NAN, &status);
  helper_test_cpublish_value_new_float_success(INFINITY, &status);
  helper_test_cpublish_value_new_float_success(-INFINITY, &status);

  cpublish_status_destroy(&status);
}

void helper_test_cpublish_value_new_string_success(const char *v,
                                                   CPublishStatus *status) {
  CPublishValue *value = cpublish_value_new_string(v);
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, status), CPublishValueTypeString);
  validate_status_ok(status);

  CPublishString result = cpublish_value_string(value, status);
  validate_status_ok(status);
  assert_string_equal(result.string, v);

  cpublish_string_destroy(&result);
  cpublish_value_destroy(value);
}

static void test_cpublish_value_new_string_success(void **state) {
  CPublishStatus status;

  helper_test_cpublish_value_new_string_success("", &status);
  helper_test_cpublish_value_new_string_success("test", &status);
  helper_test_cpublish_value_new_string_success(
      "test test test test test test test test test", &status);
  helper_test_cpublish_value_new_string_success("12345", &status);

  cpublish_status_destroy(&status);
}

static void test_cpublish_value_new_array_success(void **state) {
  CPublishStatus status;

  CPublishValue *value = cpublish_value_new_array();
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, &status), CPublishValueTypeArray);
  validate_status_ok(&status);

  // No items
  assert_int_equal(cpublish_value_array_len(value, &status), 0);
  validate_status_ok(&status);

  // Added item
  CPublishValue *item = cpublish_value_new_none();
  cpublish_value_array_push(value, item, &status);
  cpublish_value_destroy(item);
  validate_status_ok(&status);
  assert_int_equal(cpublish_value_array_len(value, &status), 1);
  validate_status_ok(&status);
  cpublish_value_array_get(value, 0, &status);
  validate_status_ok(&status);
  CPublishValueIterArray *iter = cpublish_value_array_iter(value, &status);
  validate_status_ok(&status);
  size_t count = 0;

  while (!cpublish_value_iter_array_is_done(iter, &status)) {
    count++;
    const CPublishValue *item = cpublish_value_iter_array_value(iter, &status);
    validate_status_ok(&status);
    assert_int_equal(cpublish_value_type(item, &status), CPublishValueTypeNone);
    validate_status_ok(&status);
    cpublish_value_iter_array_next(iter, &status);
    validate_status_ok(&status);
  }

  assert_int_equal(count, 1);

  cpublish_value_iter_array_destroy(iter);

  cpublish_value_destroy(value);
  cpublish_status_destroy(&status);
}

static void test_cpublish_value_new_array_with_capacity_success(void **state) {
  CPublishStatus status;

  CPublishValue *value = cpublish_value_new_array_with_capacity(1);
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, &status), CPublishValueTypeArray);
  validate_status_ok(&status);

  // No items
  assert_int_equal(cpublish_value_array_len(value, &status), 0);
  validate_status_ok(&status);

  // Added item
  CPublishValue *item = cpublish_value_new_none();
  cpublish_value_array_push(value, item, &status);
  cpublish_value_destroy(item);
  validate_status_ok(&status);
  assert_int_equal(cpublish_value_array_len(value, &status), 1);
  validate_status_ok(&status);
  cpublish_value_array_get(value, 0, &status);
  validate_status_ok(&status);
  CPublishValueIterArray *iter = cpublish_value_array_iter(value, &status);
  validate_status_ok(&status);
  size_t count = 0;

  while (!cpublish_value_iter_array_is_done(iter, &status)) {
    count++;
    const CPublishValue *item = cpublish_value_iter_array_value(iter, &status);
    validate_status_ok(&status);
    assert_int_equal(cpublish_value_type(item, &status), CPublishValueTypeNone);
    validate_status_ok(&status);
    cpublish_value_iter_array_next(iter, &status);
    validate_status_ok(&status);
  }

  assert_int_equal(count, 1);

  cpublish_value_iter_array_destroy(iter);

  cpublish_value_destroy(value);
  cpublish_status_destroy(&status);
}

static void test_cpublish_value_new_object_success(void **state) {
  CPublishStatus status;

  CPublishValue *value = cpublish_value_new_object();
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, &status),
                   CPublishValueTypeObject);
  validate_status_ok(&status);

  // No items
  assert_int_equal(cpublish_value_object_len(value, &status), 0);
  validate_status_ok(&status);

  // Added item
  CPublishValue *item = cpublish_value_new_none();
  cpublish_value_object_insert(value, "test", item, &status);
  cpublish_value_destroy(item);
  validate_status_ok(&status);
  assert_int_equal(cpublish_value_object_len(value, &status), 1);
  validate_status_ok(&status);
  cpublish_value_object_get(value, "test", &status);
  validate_status_ok(&status);
  CPublishValueIterObject *iter = cpublish_value_object_iter(value, &status);
  validate_status_ok(&status);
  size_t count = 0;

  while (!cpublish_value_iter_object_is_done(iter, &status)) {
    count++;
    const CPublishValue *item = cpublish_value_iter_object_value(iter, &status);
    validate_status_ok(&status);
    assert_int_equal(cpublish_value_type(item, &status), CPublishValueTypeNone);
    validate_status_ok(&status);
    cpublish_value_iter_object_next(iter, &status);
    validate_status_ok(&status);
  }

  assert_int_equal(count, 1);

  cpublish_value_iter_object_destroy(iter);

  cpublish_value_destroy(value);
  cpublish_status_destroy(&status);
}

static void test_cpublish_value_new_object_with_capacity_success(void **state) {
  CPublishStatus status;

  CPublishValue *value = cpublish_value_new_object_with_capacity(1);
  assert_non_null(value);

  assert_int_equal(cpublish_value_type(value, &status),
                   CPublishValueTypeObject);
  validate_status_ok(&status);

  // No items
  assert_int_equal(cpublish_value_object_len(value, &status), 0);
  validate_status_ok(&status);

  // Added item
  CPublishValue *item = cpublish_value_new_none();
  cpublish_value_object_insert(value, "test", item, &status);
  cpublish_value_destroy(item);
  validate_status_ok(&status);
  assert_int_equal(cpublish_value_object_len(value, &status), 1);
  validate_status_ok(&status);
  cpublish_value_object_get(value, "test", &status);
  validate_status_ok(&status);
  CPublishValueIterObject *iter = cpublish_value_object_iter(value, &status);
  validate_status_ok(&status);
  size_t count = 0;

  while (!cpublish_value_iter_object_is_done(iter, &status)) {
    count++;
    const CPublishValue *item = cpublish_value_iter_object_value(iter, &status);
    validate_status_ok(&status);
    assert_int_equal(cpublish_value_type(item, &status), CPublishValueTypeNone);
    validate_status_ok(&status);
    cpublish_value_iter_object_next(iter, &status);
    validate_status_ok(&status);
  }

  assert_int_equal(count, 1);

  cpublish_value_iter_object_destroy(iter);

  cpublish_value_destroy(value);
  cpublish_status_destroy(&status);
}

int main(void) {
  const struct CMUnitTest tests[] = {
      cmocka_unit_test(test_cpublish_value_new_none_success),
      cmocka_unit_test(test_cpublish_value_new_bool_success),
      cmocka_unit_test(test_cpublish_value_new_int_success),
      cmocka_unit_test(test_cpublish_value_new_float_success),
      cmocka_unit_test(test_cpublish_value_new_string_success),
      cmocka_unit_test(test_cpublish_value_new_array_success),
      cmocka_unit_test(test_cpublish_value_new_array_with_capacity_success),
      cmocka_unit_test(test_cpublish_value_new_object_success),
      cmocka_unit_test(test_cpublish_value_new_object_with_capacity_success),
  };

  return cmocka_run_group_tests(tests, NULL, NULL);
}

File: bindings/c/tests/test_value.rs
use std::ffi::{CStr, CString};

use proptest::prelude::*;

use cpublish::*;

fn arb_value_array() -> impl Strategy<Value = publish::Value> {
    prop::collection::vec(arb_value(), 0..10)
        .prop_map(publish::Value::Array)
        .boxed()
}

fn arb_value_object() -> impl Strategy<Value = publish::Value> {
    prop::collection::hash_map("[^\0]*", arb_value(), 0..10)
        .prop_map(publish::Value::Object)
        .boxed()
}

fn arb_value() -> impl Strategy<Value = publish::Value> {
    let leaf = prop_oneof![
        Just(publish::Value::None),
        any::<bool>().prop_map(publish::Value::Boolean),
        any::<i64>().prop_map(publish::Value::Integer),
        any::<f64>().prop_map(publish::Value::Float),
        "[^\0]*".prop_map(publish::Value::String),
    ];

    leaf.prop_recursive(8, 256, 10, |inner| {
        prop_oneof![
            prop::collection::vec(inner.clone(), 0..10).prop_map(publish::Value::Array),
            prop::collection::hash_map("[^\0]*", inner, 0..10).prop_map(publish::Value::Object),
        ]
    })
}

#[test]
fn test_cpublish_value_new_none_success() {
    unsafe {
        let value = cpublish_value_new_none();
        let mut status = CPublishStatus::new_ok();

        assert!(!value.is_null());
        assert_eq!(
            cpublish_value_type(value, &mut status),
            CPublishValueType::CPublishValueTypeNone
        );
        assert_eq!(
            status.status,
            CPublishStatusType::CPublishStatusTypeOk,
            "Err: {}",
            CStr::from_ptr(status.message).to_string_lossy()
        );
    }
}

proptest! {
    #[test]
    fn test_cpublish_value_new_bool_success(v: bool) {
        unsafe {
            let value = cpublish_value_new_bool(v);

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeBoolean);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            assert_eq!(cpublish_value_bool(value, &mut status), v);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
        }
    }

    #[test]
    fn test_cpublish_value_new_int_success(v: i64) {
        unsafe {
            let value = cpublish_value_new_int(v);

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeInteger);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            assert_eq!(cpublish_value_int(value, &mut status), v);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
        }
    }

    #[test]
    fn test_cpublish_value_new_float_success(v: f64) {
        unsafe {
            let value = cpublish_value_new_float(v);

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeFloat);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            if v.is_nan() {
                assert!(cpublish_value_float(value, &mut status).is_nan());
            } else {
                assert_eq!(cpublish_value_float(value, &mut status), v);
            }

            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
        }
    }

    #[test]
    fn test_cpublish_value_new_string_success(v in "[^\0]*") {
        let c_v = CString::new(v).unwrap().into_raw();

        unsafe {
            let value = cpublish_value_new_string(c_v);

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeString);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            assert_eq!(CStr::from_ptr(cpublish_value_string(value, &mut status).string), CStr::from_ptr(c_v));
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            drop(CString::from_raw(c_v));
        }
    }

    #[test]
    fn test_cpublish_value_new_array_success(v in arb_value_array()) {
        let arr = match &v {
            publish::Value::Array(arr) => arr,
            _ => panic!(),
        };

        unsafe {
            let value = cpublish_value_new_array();

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeArray);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for item in arr {
                cpublish_value_array_push(value, &CPublishValue::from(item.clone()), &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            assert_eq!(cpublish_value_array_len(value, &mut status), arr.len());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for index in 0..arr.len() {
                let item = cpublish_value_array_get(value, index, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!item.is_null());
                assert_eq!(&(*item).value, &arr[index]);
            }

            let iter = cpublish_value_array_iter(value, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            let mut index = 0;

            while !cpublish_value_iter_array_is_done(iter, &mut status) {
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(index <= arr.len());

                let item = cpublish_value_iter_array_value(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!item.is_null());
                assert_eq!(&(*item).value, &arr[index]);
                index += 1;
                cpublish_value_iter_array_next(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            assert_eq!(index, arr.len());

            cpublish_value_iter_array_destroy(iter);
        }
    }

    #[test]
    fn test_cpublish_value_new_array_with_capacity_success(v in arb_value_array()) {
        let arr = match &v {
            publish::Value::Array(arr) => arr,
            _ => panic!(),
        };

        unsafe {
            let value = cpublish_value_new_array_with_capacity(arr.len());

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeArray);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for item in arr {
                cpublish_value_array_push(value, &CPublishValue::from(item.clone()), &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            assert_eq!(cpublish_value_array_len(value, &mut status), arr.len());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for index in 0..arr.len() {
                let item = cpublish_value_array_get(value, index, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!item.is_null());
                assert_eq!(&(*item).value, &arr[index]);
            }

            let iter = cpublish_value_array_iter(value, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            let mut index = 0;

            while !cpublish_value_iter_array_is_done(iter, &mut status) {
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(index <= arr.len());

                let item = cpublish_value_iter_array_value(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!item.is_null());
                assert_eq!(&(*item).value, &arr[index]);
                index += 1;

                cpublish_value_iter_array_next(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            assert_eq!(index, arr.len());

            cpublish_value_iter_array_destroy(iter);
        }
    }

    #[test]
    fn test_cpublish_value_new_object_success(v in arb_value_object()) {
        let obj = match &v {
            publish::Value::Object(obj) => obj,
            _ => panic!(),
        };

        unsafe {
            let value = cpublish_value_new_object();

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeObject);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for (key, item) in obj {
                let key = CString::new(key.clone()).unwrap().into_raw();
                cpublish_value_object_insert(value, key, &CPublishValue::from(item.clone()), &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                drop(CString::from_raw(key));
            }

            assert_eq!(cpublish_value_object_len(value, &mut status), obj.len());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for (key, item) in obj {
                let c_key = CString::new(key.clone()).unwrap().into_raw();
                let value_item = cpublish_value_object_get(value, c_key, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!value_item.is_null());
                assert_eq!(&(*value_item).value, item);
                drop(CString::from_raw(c_key));
            }

            let iter = cpublish_value_object_iter(value, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            let mut index = 0;

            while !cpublish_value_iter_object_is_done(iter, &mut status) {
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(index <= obj.len());


                let key = cpublish_value_iter_object_key(iter, &mut status);
                assert!(!key.string.is_null());
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

                let item = cpublish_value_iter_object_value(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!item.is_null());
                assert_eq!(&(*item).value, &obj[CStr::from_ptr(key.string).to_string_lossy().as_ref()]);
                index += 1;

                cpublish_value_iter_object_next(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            assert_eq!(index, obj.len());

            cpublish_value_iter_object_destroy(iter);
        }
    }

    #[test]
    fn test_cpublish_value_new_object_with_capacity_success(v in arb_value_object()) {
        let obj = match &v {
            publish::Value::Object(obj) => obj,
            _ => panic!(),
        };

        unsafe {
            let value = cpublish_value_new_object_with_capacity(obj.len());

            assert!(!value.is_null());
            let mut status = CPublishStatus::new_ok();
            assert_eq!(cpublish_value_type(value, &mut status), CPublishValueType::CPublishValueTypeObject);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for (key, item) in obj {
                let key = CString::new(key.clone()).unwrap().into_raw();
                cpublish_value_object_insert(value, key, &CPublishValue::from(item.clone()), &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                drop(CString::from_raw(key));
            }

            assert_eq!(cpublish_value_object_len(value, &mut status), obj.len());
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            for (key, item) in obj {
                let c_key = CString::new(key.clone()).unwrap().into_raw();
                let value_item = cpublish_value_object_get(value, c_key, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!value_item.is_null());
                assert_eq!(&(*value_item).value, item);
                drop(CString::from_raw(c_key));
            }

            let iter = cpublish_value_object_iter(value, &mut status);
            assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

            let mut index = 0;

            while !cpublish_value_iter_object_is_done(iter, &mut status) {
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(index <= obj.len());

                let key = cpublish_value_iter_object_key(iter, &mut status);
                assert!(!key.string.is_null());
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());

                let item = cpublish_value_iter_object_value(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
                assert!(!item.is_null());
                assert_eq!(&(*item).value, &obj[CStr::from_ptr(key.string).to_string_lossy().as_ref()]);
                index += 1;

                cpublish_value_iter_object_next(iter, &mut status);
                assert_eq!(status.status, CPublishStatusType::CPublishStatusTypeOk, "Err: {}", CStr::from_ptr(status.message).to_string_lossy());
            }

            assert_eq!(index, obj.len());

            cpublish_value_iter_object_destroy(iter);
        }
    }
}
